Sources/Crypto/Digests/BoringSSL/Digest_boring.swift

```diff
@@ -67,81 +141,44 @@ struct OpenSSLDigestImpl<H: BoringSSLBackedHashFunction> {
     }
 
     internal func finalize() -> H.Digest {
-        // To have a non-destructive finalize operation we must allocate.
-        let copyContext = DigestContext(copying: self.context)
-        let digestBytes = copyContext.finalize()
-        return digestBytes.withUnsafeBytes {
+        return self.context.finalize {
             // We force unwrap here because if the digest size is wrong it's an internal error.
             H.Digest(bufferPointer: $0)!
         }
     }
 }
 
-class DigestContext {
-    private var contextPointer: UnsafeMutablePointer<EVP_MD_CTX>
+final class DigestContext<H: BoringSSLBackedHashFunction> {
+    @exclusivity(unchecked) private var context: H.Context
```

Comment:
The performance impact is not very high: about 1.15x in my use case. I'll remove the `@exclusivity` attribute.

I've done a little bit of testing. I'm able to convince the optimizer to eliminate the exclusivity checks when the following conditions are both satisfied.
1. `self.context` is not accessed as an `inout` variable.
2. `self.context` is not mutated in `deinit`.

The first condition makes sense, as the optimizer probably doesn't know that the `DigestContext` is uniquely referenced when `update` is called and must therefore assume that `CCryptoBoringSSL_*_Update` could access the context a second time using global memory.

The second condition is more puzzling to me. I don't understand how mutating the context in `deinit` could trigger an exclusivity violation, as we know for sure that `self` isn't referenced anywhere anymore.

If we decided to let go of zeroization and remove the `deinit` method, we could rewrite the `update` method like this to eliminate the exclusivity checks:
```swift
var copyContext = self.context
H.update(&copyContext, data: data)
self.context = copyContext
```
Despite the extra copies, it is faster than checking for exclusive access.

Reply:
Ok, I think to avoid blocking this on that performance wrinkle we should go without the exclusivity change for now. Can you file a separate Github issue that tracks eliminating that exclusivity check, ideally with a patch that we can apply that shows the two sides? I'll then see if I can work out what's happening and come up with a way to fix it.